Library/Homebrew/utils/gems.rb

```diff
@@ -191,9 +191,11 @@ def user_gem_groups
   end
 
   def write_user_gem_groups(groups)
+    return if @bundle_installed_groups == groups && GEM_GROUPS_FILE.exist?
```

Comment:
Note that this write operation is already in a `@bundle_installed_groups != groups` check so this isn't really doing anything.

You're maybe wanting to `@user_gem_groups`

Reply:
And we should also do a `@user_gem_groups = groups` at the end of this and `@user_gem_groups = []` in `forget_user_gem_groups!`